### ISSUE TITLE
RUF059 and F841 not detected when calling locals() before unpacking a tuple

### ISSUE THREAD

**Issue by btjanaka** [astral-sh/ruff#26522]
### Summary

Hi! I noticed that when I call locals() before unpacking a tuple, ruff fails to report an unused variable from unpacking the tuple. For example, in the following file:

```python
G = (1, 2)


def ruf059_with_locals():
    locals()
    a, b = G  # RUF059 not detected (false negative).


def ruf059_without_locals():
    a, b = G  # RUF059 detected (correct).


def f841_with_locals():
    locals()
    a, b = (1, 2)  # F841 not detected (false negative).


def f841_without_locals():
    a, b = (1, 2)  # F841 detected (correct).
```

The function `ruf059_with_locals` does not have any detection of `RUF059`, while `ruf059_without_locals` does have the detection. Similarly, `f841_with_locals` does not have any detection of `F841`, while `f841_without_locals` does.

For completeness, this is the output when I run `ruff check` on the above code:

```text
RUF059 Unpacked variable `a` is never used
  --> test.py:10:5
   |
 9 | def ruf059_without_locals():
10 |     a, b = G  # RUF059 not detected (correct).
   |     ^
   |
help: Prefix it with an underscore or any other dummy variable pattern

RUF059 Unpacked variable `b` is never used
  --> test.py:10:8
   |
 9 | def ruf059_without_locals():
10 |     a, b = G  # RUF059 not detected (correct).
   |        ^
   |
help: Prefix it with an underscore or any other dummy variable pattern

F841 Local variable `a` is assigned to but never used
  --> test.py:19:5
   |
18 | def f841_without_locals():
19 |     a, b = (1, 2)  # F841 detected (correct).
   |     ^
   |
help: Remove assignment to unused variable `a`

F841 Local variable `b` is assigned to but never used
  --> test.py:19:8
   |
18 | def f841_without_locals():
19 |     a, b = (1, 2)  # F841 detected (correct).
   |        ^
   |
help: Remove assignment to unused variable `b`

Found 4 errors.
```

### Version

ruff 0.15.20

**Comment by MichaReiser** [comment]
This was intentionally changed in https://github.com/astral-sh/ruff/issues/220

Arguably, in your case, `a`, and `b` are not unused. `locals` references all local symbols, including `a`, and `b`. If the result of `locals` itself is unused, it would then probably be fine to flag the variables, but I'm not sure adding the complexity to detect this is worth it (if so, Ruff should flag that the `locals` call is unused).

So I think this is intentional. 

Could you share the specific case where you ran into this? This might help inform our decision making

**Comment by btjanaka** [comment]
Thanks for the reply @MichaReiser! For a bit more context, my use case is one where I have a bunch of hyperparameters passed into a main function, and then I save the hyperparameters by calling locals(), and then call a training loop where the tuple unpacking is done from a (PyTorch or Grain) data loader:

```python
def main(param1, param2):
    all_params = locals()
    # Save all_params to JSON, PKL, or YAML file.

    for batch in dataloader:
        data1, data2 = batch  # No detection by ruff.

```

Also worth mentioning, my impression of locals() is that it would not capture variables that are declared after it is called, so a and b (or here data1 and data2) are indeed unused?

I totally understand if this is a pretty rare use case; happy to live with it :D

**Comment by MichaReiser** [comment]
Thank you, this is helpful.

> Also worth mentioning, my impression of locals() is that it would not capture variables that are declared after it is called, so a and b (or here data1 and data2) are indeed unused?



Right. I'd be open to change the rule to only disable it for all variables up to the first `locals` call in a function. Now, things get a bit tricky once control flow is involved, especially with loops.
